### PR TITLE
Adds AdamW optimiser

### LINKED ISSUE #295: Add AdamW Optimiser
### Feature description

Add an AdamW implementation to PyBOP. To achieve this, an implementation that matches the ask-tell interface should be completed. This optimiser will likely be the first pure PyBOP optimisation method.

### Motivation

[AdamW](https://arxiv.org/abs/1711.05101) provides a weight decay implementation of the Adam algorithm. A pytorch implementation can be found here: https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

This PR adds an [AdamW](https://arxiv.org/abs/1711.05101) optimiser to PyBOP. This implementation extends the Pints' Adam class and reimplements it in PyBOP. Adding this upstream to Pints' should also be considered.

This PR should be merged after #255 as it will need to be updated for the new optimiser architecture implemented.

## Issue reference
Fixes #295 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [X] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: pybop/optimisers/pints_optimisers.py [hunk 1]
```diff
@@ -60,10 +62,43 @@ class Adam(BasePintsOptimiser):
...
+    optimiser that incorporates weight decay. AdamW is designed to be more
+    robust and stable for training deep neural networks, particularly when
+    using larger learning rates.
+
+    Parameters
+    ----------
+    **optimiser_kwargs : optional
+        Valid PyBOP option keys and their values, for example:
+        x0 : array_like
+            Initial position from which optimization will start.
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:89]:
```suggestion
            Initial position from which optimisation will start.
```

File: pybop/optimisers/pints_optimisers.py [hunk 2]
```diff
@@ -60,10 +62,43 @@ class Adam(BasePintsOptimiser):
...
+        stacklevel=2,
+    )
+
     def __init__(self, cost, **optimiser_kwargs):
         super().__init__(cost, PintsAdam, **optimiser_kwargs)
 
 
+class AdamW(BasePintsOptimiser):
+    """
+    Implements the AdamW optimization algorithm in PyBOP.
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:77]:
```suggestion
    Implements the AdamW optimisation algorithm in PyBOP.
```

File: pybop/optimisers/_adamw.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
+#
+# Reimplements the Pints' Adam Class with a Weight Decay addition
```
Comment (BradyPlanden) [inline pybop/optimisers/_adamw.py:2]:
```suggestion
# Extends the Pints' Adam Class with a Weight Decay addition
```

**Review by NicolaCourtier — APPROVED** [review]
Looks great! Thanks very much @BradyPlanden for the nice addition of the AdamW optimiser. Please update the table in the Readme from Adam to AdamW, otherwise good to go.
